### PR TITLE
found a bug in code where AI can change client gamestate directly

### PR DESCRIPTION
**Pull request by nullkiller**
There is an issue detected in Battle AI code. It can change client gamestate which in turn can crash the whole game.
Made a clone of the bonus to avoid gamestate modification

### CODE REVIEW

File: AI/BattleAI/BattleAI.cpp [hunk 1]
```diff
@@ -578,6 +578,7 @@ void CBattleAI::attemptCastingSpell()
 		size_t minTurnSpan = ourUnits/3; //todo: tweak this
 
 		std::vector<battle::Units> newTurnOrder;
+
 		state.battleGetTurnOrder(newTurnOrder, amount, 2);
```
Comment (nullkiller) [inline AI/BattleAI/BattleAI.cpp:582]: This thing runs in parallel and invokes std::sort on a list of stacks with CMP_Stack comparator

File: AI/BattleAI/BattleAI.cpp [hunk 2]
```diff
@@ -621,7 +622,7 @@ void CBattleAI::attemptCastingSpell()
 	for(PossibleSpellcast & psc : possibleCasts)
 		tasks.push_back(std::bind(evaluateSpellcast, &psc));
 
-	uint32_t threadCount = boost::thread::hardware_concurrency();
+	uint32_t threadCount = 30 + boost::thread::hardware_concurrency();
```
Comment (nullkiller) [inline AI/BattleAI/BattleAI.cpp:625]: Will revert. Needed to reproduce the issue

File: AI/BattleAI/StackWithBonuses.cpp [hunk 3]
```diff
@@ -18,7 +18,8 @@ void actualizeEffect(TBonusListPtr target, const Bonus & ef)
 	{
 		if(bonus->source == Bonus::SPELL_EFFECT && bonus->type == ef.type && bonus->subtype == ef.subtype)
 		{
-			bonus->turnsRemain = std::max(bonus->turnsRemain, ef.turnsRemain);
+			//TODO: this code changes gamestate
+			//bonus->turnsRemain = std::max(bonus->turnsRemain, ef.turnsRemain);
```
Comment (nullkiller) [inline AI/BattleAI/StackWithBonuses.cpp:22]: And here is what causes this issue. Bonuses from TBonusListPtr target can be from real stacks from gamestate.

File: lib/battle/BattleInfo.cpp [hunk 4]
```diff
@@ -1066,33 +1066,21 @@ bool CMP_stack::operator()(const battle::Unit * a, const battle::Unit * b)
 	{
 	case 0: //catapult moves after turrets
 		return a->creatureIndex() > b->creatureIndex(); //catapult is 145 and turrets are 149
-	case 1: //fastest first, upper slot first
```
Comment (nullkiller) [inline lib/battle/BattleInfo.cpp:1069]: The logic of phase 1 and phase 3 looks absolutely the same

File: lib/battle/BattleInfo.cpp [hunk 5]
```diff
@@ -1066,33 +1066,21 @@ bool CMP_stack::operator()(const battle::Unit * a, const battle::Unit * b)
...
+			
 			if(as != bs)
 				return as > bs;
 
 			if(a->unitSide() == b->unitSide())
 				return a->unitSlot() < b->unitSlot();
-			else if (a->unitSide() == side || b->unitSide() == side)
-				return a->unitSide() != side;
-
-			return std::addressof(a) > std::addressof(b);
```
Comment (nullkiller) [inline lib/battle/BattleInfo.cpp:1095]: This does not cause issues but still looks inconsistent and was my first guess. this->side can sometimes be 255 or something and it looks like usual expected scenario.

File: AI/BattleAI/StackWithBonuses.cpp (hunk 3 above)
Comment (nullkiller) [reply AI/BattleAI/StackWithBonuses.cpp:22]: After  commenting out this code I was not able to reproduce the crash again.

Comment (nullkiller) [reply AI/BattleAI/StackWithBonuses.cpp]: Cloned the bonus from the gamestate and changed the clone

**Review by alexvins — APPROVED** [review]

**Pull request merged**